### ISSUE TITLE
Pushing an empty UrlList has no effect

### ISSUE THREAD

**Issue by blubber** [googleads/googleads-python-lib#63]
We are currently migrating our product to the new destination URLs and are hitting a snag. Setting finalUrls and finalMobilUrls on regular keywords works fine, but after setting URLs it is impossible to clear the list. This is the dictionary that we are passing to the mutate method:

    {'operand': {'adGroupId': '<redacted>',
              'criterion': {'id': '<redacted>',
                            'matchType': u'EXACT',
                            'text': u'some keyword',
                            'xsi_type': 'Keyword'},
              'destinationUrl': '',
              'finalMobileUrls': {'urls': []},
              'finalUrls': {'urls': []},
              'xsi_type': 'BiddableAdGroupCriterion'},
  'operator': 'SET'} 

The following XML blob is created by suds:

    <?xml version="1.0" ?>
    <SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns0="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="https://adwords.google.com/api/adwords/cm/v201502" xmlns:tns="https://adwords.google.com/api/adwords/cm/v201502" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
        <SOAP-ENV:Header>
            <tns:RequestHeader>
                <tns:clientCustomerId>REDACTED</tns:clientCustomerId>
                <tns:developerToken>REDACTED</tns:developerToken>
                <tns:userAgent>(AwApi-Python, googleads/3.4.1, Python/2.7)</tns:userAgent>
                <tns:validateOnly>false</tns:validateOnly>
                <tns:partialFailure>false</tns:partialFailure>
            </tns:RequestHeader>
        </SOAP-ENV:Header>
        <ns0:Body>
            <ns1:mutate>
                <ns1:operations>
                    <ns1:operator>SET</ns1:operator>
                    <ns1:operand xsi:type="ns1:BiddableAdGroupCriterion">
                        <ns1:adGroupId>REDACTED</ns1:adGroupId>
                        <ns1:criterion xsi:type="ns1:Keyword">
                            <ns1:id>REDACTED</ns1:id>
                            <ns1:Criterion.Type>Keyword</ns1:Criterion.Type>
                            <ns1:text>some keyword</ns1:text>
                            <ns1:matchType>EXACT</ns1:matchType>
                        </ns1:criterion>
                        <ns1:AdGroupCriterion.Type>BiddableAdGroupCriterion</ns1:AdGroupCriterion.Type>
                        <ns1:destinationUrl/>
                    </ns1:operand>
                </ns1:operations>
            </ns1:mutate>
        </ns0:Body>
    </SOAP-ENV:Envelope>

In this XML blob the new (empty) UrlLists for finalUrls and finalMobileUrls are missing, and hence they are not cleared in AdWords. Setting the urls key to None or the finalUrls to {} or None doesn't solve the problem. It appears that either the adwords client or suds is removing the empty object.

Edit: this is using the latest versions of suds and googleads on Python 2.7.

**Comment by msaniscalchi** [comment]
Thanks for the report, we'll look into it.

**Comment by ddofborg** [comment]
Hello! Any updates on this one?

**Comment by ddofborg** [comment]
@msaniscalchi, did you look into this error? We are having the same problem.

**Comment by adorsey** [comment]
I've also been having problems with this for a while too, and have traced it down to a class in the suds-jurko library code.
suds-jurko/mx/appender.py
```python
class ObjectAppender(Appender):
    """
    An L{Object} appender.
    """

    def append(self, parent, content):
        object = content.value
        if self.optional(content) and footprint(object) == 0:
            return
        child = self.node(content)
        parent.append(child)
        for item in object:
            cont = Content(tag=item[0], value=item[1])
            Appender.append(self, child, cont)
```

Let’s say you’re sending an operand which contains:
```
AdGroupCriterionOperation: (AdGroupCriterionOperation){
   operator = "SET"
   Operation.Type = "AdGroupCriterionOperation"
   operand = 
      (BiddableAdGroupCriterion){
         adGroupId = "1234567890"
         criterion = 
            (Keyword){
               id = "1234567890"
               type = "i forget what the legal values are here but it's irrelevant"
               Criterion.Type = "Keyword"
               text = "some text"
               matchType = "Broad"
            }
         # some more stuff
         finalUrls = 
         {
              urls[] = <empty> 
          } # notice how this isn't formatted with a (type) like the 'criterion' has (Keyword)
         # some more stuff
```

In ObjectAppender.append, the check:
```python
        if self.optional(content) and footprint(object) == 0:
            return
```
is evaluating to `True` and returning before the finalUrls object can be built because the 'footprint' is zero.  I expect this doesn't happen on TextAds because they're a required field (but maybe it happens for finalMobileUrls?), so the 'optional' part of the check is preventing the bare return.

The footprint function, for completeness:
```python
def footprint(sobject):
    """
    Get the I{virtual footprint} of the object.
    This is really a count of the attributes in the branch with a significant value.
    @param sobject: A suds object.
    @type sobject: L{Object}
    @return: The branch footprint.
    @rtype: int
    """
    n = 0
    for a in sobject.__keylist__:
        v = getattr(sobject, a)
        if v is None: continue
        if isinstance(v, Object):
            n += footprint(v)
            continue
        if hasattr(v, '__len__'):
            if len(v): n += 1
            continue
        n +=1
    return n
```

I'm not sure what a good fix would be, but it seems like it's in the layer of this library and NOT a suds issue; I've definitely sent empty lists to optional fields before (final URLs in feed items, for example).  The lack of type on the finalUrls object seems odd, when __repr__-ed it looks like this so maybe it's okay:
```
UrlList: (UrlList){
   urls[] = <empty>
 }<empty>
```

What do you all think?  @msaniscalchi, I'm especially curious for your thoughts?  I'd really like to patch my app so that my users can send these changes to the API.

I'll see if there's anything I can do to get around this and post here if I have any luck.

**Comment by adorsey** [comment]
After digging some more, I'm not convinced that I've sent empty lists before.  In feed items, you can clear a URL list by sending an attribute ID with no value in it.  TextAds are immune to this because any change to a URL turns into a remove/recreate operation of the entire entity, and if no URL is included then it will never be set on it.

I tried several things around setting xsi_type into the operand before passing it to the client library, but have not been able to get this to work.

**Comment by adorsey** [comment]
I've been able to get around this issue by writing my own 'footprint' function over the one in suds at server startup.  It's stood up to my testing, but obviously isn't anywhere near ideal; use at your own risk!

```python
EMPTY_LIST_PROPERTY_STRINGS = frozenset((
                                         'urls',
                                         'appUrls',
                                         'finalUrls',
                                         'finalMobileUrls',
                                         'finalAppUrls',
                                         ))

def adwords_empty_final_url_footprint(sobject):
    """
        Get the I{virtual footprint} of the object.
        This is really a count of the attributes in the branch with a significant value.
        @param sobject: A suds object.
        @type sobject: L{Object}
        @return: The branch footprint.
        @rtype: int
    """
    n = 0
    for a in sobject.__keylist__:
        v = getattr(sobject, a)
        if v is None: continue
        if isinstance(v, Object):
            n += adwords_empty_final_url_footprint(v)
            continue
        if hasattr(v, '__len__'):
            if len(v): n += 1
            # START HACKED IN CODE
            elif a in EMPTY_LIST_PROPERTY_STRINGS:
                n += 1
            # END HACKED IN CODE
            continue
        n += 1
    return n

# BE CAREFUL, the way you'll want to do the following step will depend on your application
# requirements; this is the absolute simplest case and assumes single-thread access of suds
# dedicated to use with adwords.
# Substitute the hacked function into the suds library
sudsobject.footprint = adwords_empty_final_url_footprint
```

**Comment by msaniscalchi** [comment]
Hello all, sorry for the delay on this. We've decided to patch the underlying issue with the suds-jurko library in lieu of a public release of the fix included in suds-jurko 0.7, which doesn't seem like it will be getting released any time soon.

The patch is anticipated to go out with a few other small fixes soon.
